File: src/steam.ts
import { existsSync } from "node:fs";
import { mkdir, readdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { Config } from "./config";

function parseVdfPairs(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const match of content.matchAll(/^\s+"(\w+)"\s+"([^"]*?)"/gm)) {
    result[match[1]!.toLowerCase()] = match[2]!;
  }
  return result;
}

export async function getLibraryPaths(): Promise<string[]> {
  const content = await Bun.file(join(Config.STEAM_ROOT, "steamapps/libraryfolders.vdf")).text();
  const paths = [join(Config.STEAM_ROOT, "steamapps")];
  for (const match of content.matchAll(/"path"\s+"([^"]+)"/g)) {
    paths.push(join(match[1]!, "steamapps"));
  }
  return [...new Set(paths)];
}

export async function buildAppIdMap(libraryPaths: string[]): Promise<Map<string, string>> {
  const map = new Map<string, string>();
  for (const libPath of libraryPaths) {
    let entries: string[];
    try {
      entries = await readdir(libPath);
    } catch {
      continue;
    }
    for (const entry of entries) {
      if (!entry.startsWith("appmanifest_") || !entry.endsWith(".acf")) continue;
      try {
        const values = parseVdfPairs(await Bun.file(join(libPath, entry)).text());
        if (values.appid && values.name) map.set(values.appid, values.name);
      } catch {
        // skip unreadable manifests
      }
    }
  }
  return map;
}

function parseShortcutsVdf(buf: Uint8Array): Map<string, string> {
  let pos = 0;
  const dec = new TextDecoder();

  function readCStr(): string {
    const s = pos;
    while (pos < buf.length && buf[pos] !== 0) pos++;
    const str = dec.decode(buf.slice(s, pos));
    pos++;
    return str;
  }

  function readU32(): number {
    const v =
      (buf[pos]! | (buf[pos + 1]! << 8) | (buf[pos + 2]! << 16) | (buf[pos + 3]! << 24)) >>> 0;
    pos += 4;
    return v;
  }

  function skip(type: number): void {
    if (type === 0x01) readCStr();
    else if (type === 0x02) pos += 4;
    else if (type === 0x00) {
      while (pos < buf.length && buf[pos] !== 0x08) {
        const t = buf[pos++]!;
        readCStr();
        skip(t);
      }
      pos++;
    }
  }

  const map = new Map<string, string>();
  if (buf[pos] !== 0x00) return map;
  pos++;
  readCStr(); // "shortcuts"

  while (pos < buf.length && buf[pos] !== 0x08) {
    const type = buf[pos++]!;
    readCStr(); // index "0", "1", …

    if (type !== 0x00) {
      skip(type);
      continue;
    }

    let appId: number | null = null;
    let appName: string | null = null;

    while (pos < buf.length && buf[pos] !== 0x08) {
      const t = buf[pos++]!;
      const field = readCStr();
      if (t === 0x02 && field.toLowerCase() === "appid") appId = readU32();
      else if (t === 0x01 && field.toLowerCase() === "appname") appName = readCStr();
      else skip(t);
    }
    pos++;

    if (appId !== null && appName) map.set(String(appId), appName);
  }
  return map;
}

export async function loadShortcuts(): Promise<Map<string, string>> {
  if (!Config.STEAM_USER_ID) return new Map();
  const accountId = Number(BigInt(Config.STEAM_USER_ID) & 0xffffffffn);
  try {
    const buf = await Bun.file(
      join(Config.STEAM_ROOT, `userdata/${accountId}/config/shortcuts.vdf`),
    ).arrayBuffer();
    return parseShortcutsVdf(new Uint8Array(buf));
  } catch {
    return new Map();
  }
}

export async function getRunningGame(
  appIdMap: Map<string, string>,
  shortcuts: Map<string, string>,
): Promise<{ name: string; appId: string } | null> {
  let pids: string[];
  try {
    pids = await readdir("/proc");
  } catch {
    return null;
  }
  for (const pid of pids) {
    if (!/^\d+$/.test(pid)) continue;
    try {
      const buf = await readFile(`/proc/${pid}/environ`).catch(() => null);
      if (!buf) continue;
      const vars = new TextDecoder().decode(buf).split("\0");
      const entry = vars.find(v => v.startsWith("SteamAppId="));
      if (!entry) continue;
      const appId = entry.slice("SteamAppId=".length);
      if (appId && appId !== "0") {
        const name = appIdMap.get(appId) ?? shortcuts.get(appId) ?? `Unknown game (appid ${appId})`;
        return { name, appId };
      }
    } catch {
      // process exited or not readable
    }
  }
  return null;
}

export async function findSteamIconUrl(appId: string): Promise<string | null> {
  const dir = join(Config.STEAM_ROOT, `appcache/librarycache/${appId}`);
  let entries: string[];
  try {
    entries = await readdir(dir);
  } catch {
    return null;
  }

  // 1. Prefer the 600×900 portrait capsule — the full library artwork Steam caches locally.
  if (entries.includes("library_600x900.jpg")) {
    return `https://cdn.cloudflare.steamstatic.com/steam/apps/${appId}/library_600x900.jpg`;
  }

  // 2. Fall back to the 616×353 landscape capsule from the store CDN.
  if (entries.length > 0) {
    return `https://cdn.cloudflare.steamstatic.com/steam/apps/${appId}/capsule_616x353.jpg`;
  }

  // 3. Last resort: small community icon hash file.
  const iconFile = entries.find(e => /^[0-9a-f]{40}\.jpg$/.test(e));
  if (iconFile) {
    return `https://cdn.cloudflare.steamstatic.com/steamcommunity/public/images/apps/${appId}/${iconFile.slice(0, -4)}.jpg`;
  }

  return null;
}

export function findShortcutIconPath(appId: string): string | null {
  if (!Config.STEAM_USER_ID) return null;
  const accountId = Number(BigInt(Config.STEAM_USER_ID) & 0xffffffffn);
  const base = join(Config.STEAM_ROOT, `userdata/${accountId}/config/grid`);
  for (const ext of ["png", "jpg", "ico"]) {
    const p = join(base, `${appId}_icon.${ext}`);
    if (existsSync(p)) return p;
  }
  return null;
}

// ── Discord detectable games ──────────────────────────────────────────────────
// Maps lowercase game name / alias → { appId, iconUrl }.
// appId: used as IPC client_id so Discord shows the correct voice-channel icon.
// iconUrl: Discord's official CDN icon for use as large_image in rich presence.

export interface DetectableGameInfo {
  appId: string;
  iconUrl: string | null;
}

const DETECTABLE_CACHE_PATH = join(
  process.env.HOME!,
  ".local/share/steam-discord/detectable-cache.json",
);
const DETECTABLE_CACHE_VERSION = 2;
const DETECTABLE_TTL_MS = 7 * 24 * 60 * 60 * 1000; // 1 week

let detectableNameMap: Map<string, DetectableGameInfo> | null = null;

async function loadDetectableMap(): Promise<Map<string, DetectableGameInfo>> {
  if (detectableNameMap) return detectableNameMap;

  // Try the on-disk cache first to avoid a network round-trip.
  try {
    const cached = JSON.parse(await readFile(DETECTABLE_CACHE_PATH, "utf8")) as {
      version: number;
      timestamp: number;
      entries: [string, DetectableGameInfo][];
    };
    if (
      cached.version === DETECTABLE_CACHE_VERSION &&
      Date.now() - cached.timestamp < DETECTABLE_TTL_MS
    ) {
      detectableNameMap = new Map(cached.entries);
      return detectableNameMap;
    }
  } catch {}

  // Fetch a fresh copy from Discord's public endpoint.
  try {
    const resp = await fetch("https://discord.com/api/v10/applications/detectable");
    if (resp.ok) {
      const list = (await resp.json()) as Array<{
        id: string;
        name: string;
        icon_hash?: string | null;
        cover_image_hash?: string | null;
        aliases?: string[];
      }>;
      detectableNameMap = new Map();
      for (const g of list) {
        // Prefer cover_image_hash (the game directory artwork, ~512px square)
        // over icon_hash (the small app icon, typically 32–64px).
        const imageHash = g.cover_image_hash ?? g.icon_hash ?? null;
        const info: DetectableGameInfo = {
          appId: g.id,
          iconUrl: imageHash
            ? `https://cdn.discordapp.com/app-icons/${g.id}/${imageHash}.png?size=512`
            : null,
        };
        detectableNameMap.set(g.name.toLowerCase(), info);
        for (const alias of g.aliases ?? []) {
          detectableNameMap.set(alias.toLowerCase(), info);
        }
      }
      try {
        await mkdir(join(process.env.HOME!, ".local/share/steam-discord"), { recursive: true });
        await writeFile(
          DETECTABLE_CACHE_PATH,
          JSON.stringify({
            version: DETECTABLE_CACHE_VERSION,
            timestamp: Date.now(),
            entries: [...detectableNameMap],
          }),
        );
      } catch {}
      return detectableNameMap;
    }
  } catch {
    console.warn("[Discord] Could not fetch detectable games list — voice icon may show '?'");
  }

  detectableNameMap ??= new Map();
  return detectableNameMap;
}

/**
 * Returns the Discord application ID and official icon URL for a game by name
 * (from Discord's detectable-games registry), or null if unrecognised.
 */
export async function findDiscordGame(gameName: string): Promise<DetectableGameInfo | null> {
  const map = await loadDetectableMap();
  return map.get(gameName.toLowerCase()) ?? null;
}

File: src/steam.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// ── Text VDF ──────────────────────────────────────────────────────────────────

var vdfPairRe = regexp.MustCompile(`(?m)^\s+"(\w+)"\s+"([^"]*?)"`)

func parseVdfPairs(content string) map[string]string {
	result := map[string]string{}
	for _, m := range vdfPairRe.FindAllStringSubmatch(content, -1) {
		result[strings.ToLower(m[1])] = m[2]
	}
	return result
}

var vdfPathRe = regexp.MustCompile(`"path"\s+"([^"]+)"`)

func getLibraryPaths() ([]string, error) {
	vdfPath := filepath.Join(Config.SteamRoot, "steamapps/libraryfolders.vdf")
	data, err := os.ReadFile(vdfPath)
	if err != nil {
		return nil, err
	}
	content := string(data)
	paths := []string{filepath.Join(Config.SteamRoot, "steamapps")}
	for _, m := range vdfPathRe.FindAllStringSubmatch(content, -1) {
		paths = append(paths, filepath.Join(m[1], "steamapps"))
	}
	return uniqueStrings(paths), nil
}

func buildAppIDMap(libraryPaths []string) map[string]string {
	m := map[string]string{}
	for _, libPath := range libraryPaths {
		entries, err := os.ReadDir(libPath)
		if err != nil {
			continue
		}
		for _, e := range entries {
			name := e.Name()
			if !strings.HasPrefix(name, "appmanifest_") || !strings.HasSuffix(name, ".acf") {
				continue
			}
			data, err := os.ReadFile(filepath.Join(libPath, name))
			if err != nil {
				continue
			}
			vals := parseVdfPairs(string(data))
			if vals["appid"] != "" && vals["name"] != "" {
				m[vals["appid"]] = vals["name"]
			}
		}
	}
	return m
}

// ── Binary shortcuts.vdf ──────────────────────────────────────────────────────

type shortcutInfo struct {
	Name string
	Icon string // may be empty
}

func parseShortcutsVdf(buf []byte) map[string]shortcutInfo {
	pos := 0
	result := map[string]shortcutInfo{}

	readCStr := func() string {
		start := pos
		for pos < len(buf) && buf[pos] != 0 {
			pos++
		}
		s := string(buf[start:pos])
		pos++ // consume null terminator
		return s
	}

	readU32 := func() uint32 {
		if pos+4 > len(buf) {
			return 0
		}
		v := binary.LittleEndian.Uint32(buf[pos:])
		pos += 4
		return v
	}

	var skip func(t byte)
	skip = func(t byte) {
		switch t {
		case 0x01:
			readCStr()
		case 0x02:
			pos += 4
		case 0x00:
			for pos < len(buf) && buf[pos] != 0x08 {
				t2 := buf[pos]
				pos++
				readCStr()
				skip(t2)
			}
			pos++ // consume 0x08
		}
	}

	if pos >= len(buf) || buf[pos] != 0x00 {
		return result
	}
	pos++
	readCStr() // "shortcuts"

	for pos < len(buf) && buf[pos] != 0x08 {
		t := buf[pos]
		pos++
		readCStr() // index "0", "1", …

		if t != 0x00 {
			skip(t)
			continue
		}

		var appID uint32
		var hasAppID bool
		var appName, icon string

		for pos < len(buf) && buf[pos] != 0x08 {
			ft := buf[pos]
			pos++
			field := strings.ToLower(readCStr())
			switch {
			case ft == 0x02 && field == "appid":
				appID = readU32()
				hasAppID = true
			case ft == 0x01 && field == "appname":
				appName = readCStr()
			case ft == 0x01 && field == "icon":
				icon = readCStr()
			default:
				skip(ft)
			}
		}
		pos++ // consume 0x08

		if hasAppID && appName != "" {
			result[fmt.Sprintf("%d", appID)] = shortcutInfo{Name: appName, Icon: icon}
		}
	}
	return result
}

func loadShortcuts() map[string]shortcutInfo {
	if Config.SteamUserID == "" {
		return map[string]shortcutInfo{}
	}
	accountID := steamAccountID(Config.SteamUserID)
	path := filepath.Join(Config.SteamRoot, fmt.Sprintf("userdata/%d/config/shortcuts.vdf", accountID))
	data, err := os.ReadFile(path)
	if err != nil {
		return map[string]shortcutInfo{}
	}
	return parseShortcutsVdf(data)
}

// steamAccountID extracts the lower 32 bits of a 64-bit Steam ID string.
func steamAccountID(steamID64 string) uint32 {
	var id uint64
	_, _ = fmt.Sscanf(steamID64, "%d", &id)
	return uint32(id & 0xffffffff)
}

// ── Process detection ─────────────────────────────────────────────────────────

var digitRe = regexp.MustCompile(`^\d+$`)

type runningGameResult struct {
	Name  string // empty = found but unknown
	AppID string
	PID   string
}

func isGameStillRunning(pid, appID string) bool {
	data, err := os.ReadFile(fmt.Sprintf("/proc/%s/environ", pid))
	if err != nil {
		return false
	}
	target := "SteamAppId=" + appID
	for _, v := range strings.Split(string(data), "\x00") {
		if v == target {
			return true
		}
	}
	return false
}

func getRunningGame(appIDMap map[string]string, shortcuts map[string]shortcutInfo) *runningGameResult {
	entries, err := os.ReadDir("/proc")
	if err != nil {
		return nil
	}
	for _, e := range entries {
		pid := e.Name()
		if !digitRe.MatchString(pid) {
			continue
		}
		data, err := os.ReadFile(fmt.Sprintf("/proc/%s/environ", pid))
		if err != nil {
			continue
		}
		var appID string
		for _, v := range strings.Split(string(data), "\x00") {
			if strings.HasPrefix(v, "SteamAppId=") {
				appID = strings.TrimPrefix(v, "SteamAppId=")
				break
			}
		}
		if appID == "" || appID == "0" {
			continue
		}
		name := appIDMap[appID]
		if name == "" {
			if sc, ok := shortcuts[appID]; ok {
				name = sc.Name
			}
		}
		return &runningGameResult{Name: name, AppID: appID, PID: pid}
	}
	return nil
}

// ── Icon resolution ───────────────────────────────────────────────────────────

var hashJpgRe = regexp.MustCompile(`^[0-9a-f]{40}\.jpg$`)

func findSteamIconPath(appID string) string {
	dir := filepath.Join(Config.SteamRoot, "appcache/librarycache", appID)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return ""
	}
	for _, e := range entries {
		if hashJpgRe.MatchString(e.Name()) {
			return filepath.Join(dir, e.Name())
		}
	}
	return ""
}

func findShortcutIconPath(appID, vdfIconPath string) string {
	// VDF icon field - exactly what Steam displays in the sidebar.
	if vdfIconPath != "" {
		if _, err := os.Stat(vdfIconPath); err == nil {
			return vdfIconPath
		}
	}
	if Config.SteamUserID == "" {
		return ""
	}
	accountID := steamAccountID(Config.SteamUserID)
	base := filepath.Join(Config.SteamRoot, fmt.Sprintf("userdata/%d/config/grid", accountID))
	for _, ext := range []string{"png", "jpg", "ico"} {
		p := filepath.Join(base, fmt.Sprintf("%s_icon.%s", appID, ext))
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}

// ── Discord detectable games ──────────────────────────────────────────────────

type detectableGameInfo struct {
	AppID   string `json:"appId"`
	IconURL string `json:"iconUrl"` // may be empty
}

type detectableCache struct {
	Version   int                           `json:"version"`
	Timestamp int64                         `json:"timestamp"`
	Entries   map[string]detectableGameInfo `json:"entries"`
}

const detectableCacheVersion = 2
const detectableTTL = 7 * 24 * time.Hour

var detectableMap map[string]detectableGameInfo // keyed by lowercase name/alias

func loadDetectableMap() map[string]detectableGameInfo {
	if detectableMap != nil {
		return detectableMap
	}

	cachePath := filepath.Join(Config.Home, ".local/share/steamd/detectable-cache.json")

	// Try on-disk cache first.
	if data, err := os.ReadFile(cachePath); err == nil {
		var c detectableCache
		if json.Unmarshal(data, &c) == nil &&
			c.Version == detectableCacheVersion &&
			time.Since(time.UnixMilli(c.Timestamp)) < detectableTTL {
			detectableMap = c.Entries
			return detectableMap
		}
	}

	// Fetch from Discord's public endpoint.
	resp, err := http.Get("https://discord.com/api/v10/applications/detectable")
	if err != nil || resp.StatusCode != 200 {
		fmt.Fprintln(os.Stderr, "[Discord] Could not fetch detectable games list - voice icon may show '?'")
		detectableMap = map[string]detectableGameInfo{}
		return detectableMap
	}
	defer resp.Body.Close()

	var list []struct {
		ID             string   `json:"id"`
		Name           string   `json:"name"`
		IconHash       string   `json:"icon_hash"`
		CoverImageHash string   `json:"cover_image_hash"`
		Aliases        []string `json:"aliases"`
	}
	if json.NewDecoder(resp.Body).Decode(&list) != nil {
		detectableMap = map[string]detectableGameInfo{}
		return detectableMap
	}

	detectableMap = make(map[string]detectableGameInfo, len(list))
	for _, g := range list {
		imageHash := g.CoverImageHash
		if imageHash == "" {
			imageHash = g.IconHash
		}
		info := detectableGameInfo{AppID: g.ID}
		if imageHash != "" {
			info.IconURL = fmt.Sprintf("https://cdn.discordapp.com/app-icons/%s/%s.png?size=512", g.ID, imageHash)
		}
		detectableMap[strings.ToLower(g.Name)] = info
		for _, alias := range g.Aliases {
			detectableMap[strings.ToLower(alias)] = info
		}
	}

	// Persist cache.
	if err := os.MkdirAll(filepath.Dir(cachePath), 0o755); err == nil {
		c := detectableCache{
			Version:   detectableCacheVersion,
			Timestamp: time.Now().UnixMilli(),
			Entries:   detectableMap,
		}
		if data, err := json.Marshal(c); err == nil {
			_ = os.WriteFile(cachePath, data, 0o644)
		}
	}
	return detectableMap
}

func findDiscordGame(gameName string) *detectableGameInfo {
	m := loadDetectableMap()
	if info, ok := m[strings.ToLower(gameName)]; ok {
		return &info
	}
	return nil
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func uniqueStrings(in []string) []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(in))
	for _, s := range in {
		if !seen[s] {
			seen[s] = true
			out = append(out, s)
		}
	}
	return out
}
